platform/mbed_toolchain.h

```diff
@@ -152,6 +152,17 @@
 #endif
 #endif
 
+/** MBED_BARRIER
+ * Stop the compiler moving memory accesses.
+ */
+#ifdef __CC_ARM
+#define MBED_BARRIER() __memory_changed()
```

Comment:
It would be nice to have a detailed description - and maybe more appropriate naming - of what `MBED_BARRIER` is; where it should be used, what it does and what it doesn't do. 

From my understanding `asm volatile("" : : : "memory")` forces the compiler to not reorder the code it produces however there is no guarantee that the code won't be reordered by the CPU or that memory writes have completed. It doesn't replaces CPU barrier instructions.